File: Skylines/clojure/skylines.clj
(defn rectLeft [rect] (nth rect 0))
(defn rectRight [rect] (nth rect 2))
(defn rectHeight [rect] (nth rect 1))

(defn make-moves [indeces heightMap lastMove]
    (if (or (nil? heightMap) (empty? heightMap))
        '()
        (let [i (first indeces)
              h (first heightMap)]
            (if (not (= h lastMove))
                (cons i (cons h (make-moves (rest indeces) (rest heightMap) h)))
                (make-moves (rest indeces) (rest heightMap) lastMove)))))

(defn skyline [inputs]
    (let [rectangles (map 
                        (fn [s]
                            (map 
                                (fn [ss] (Integer/valueOf ss))
                                (clojure.string/split (subs s 1 (- (.length s) 1)) #",")))
                        (clojure.string/split inputs #" "))
          heightMap 
            (map (fn [i]
                    (let [heights (map (fn [r] (rectHeight r)) (filter (fn [r] (and (>= i (rectLeft r)) (< i (rectRight r)))) rectangles))]
                        (if (or (nil? heights) (empty? heights))
                            0
                            (apply max heights))))
                 (take (+ 1 (apply max (map (fn [r] (rectRight r)) rectangles))) (iterate inc 0)))]
        (clojure.string/join "," (make-moves (take (count heightMap) (iterate inc 0)) heightMap 0))))

(doseq [inputs (repeatedly read-line) :while inputs] (println (skyline inputs)))

File: Skylines/java/src/main/java/Skylines.java
import java.util.*;
import java.util.stream.Collectors;

public class Skylines
{
    public static void main(String... args)
    {
        Scanner stdin = new Scanner(System.in);
        String line;
        while (stdin.hasNextLine() && (line = stdin.nextLine()) != null)
        {
            String[] inputs = line.split(" ");
            List<Rectangle> rectangles = Arrays.asList(inputs).stream().map(input -> {
                String[] points = input.split(",");
                points[0] = points[0].substring(1);
                points[2] = points[2].substring(0, points[2].length() - 1);
                return new Rectangle(points);
            }).collect(Collectors.toList());

            List<Integer> heightMap = new ArrayList<>(Collections.nCopies(rectangles.stream().map(Rectangle::getRight).max(Comparator.<Integer>naturalOrder()).get() + 1, 0));
            for (Rectangle rect : rectangles)
            {
                for (int i = rect.getLeft(); i < rect.getRight(); i++)
                {
                    if (heightMap.get(i) < rect.getHeight())
                    {
                        heightMap.set(i, rect.getHeight());
                    }
                }
            }

            List<String> moves = new ArrayList<>();
            Integer last = 0;
            int heightMapSize = heightMap.size();
            for (int i = 0; i < heightMapSize; i++)
            {
                int height = heightMap.get(i);
                if (height != last)
                {
                    moves.add(String.valueOf(i));
                    moves.add(String.valueOf(height));
                }
                last = height;
            }
            System.out.println(String.join(",", moves));
        }
    }
}